src/cli/generate_dictionary.rs

```diff
@@ -0,0 +1,127 @@
+//! Generate a static prediction table from a sample of JS source files.
+
+use ::generic::FromJSON;
+use ::io::entropy::dictionary::{Dictionary, DictionaryBuilder, FilesContaining};
+use ::io::statistics::Instances;
+use ::io::{Path as IOPath, TokenSerializer};
+use ::source::{Shift, SourceParser};
+use ::specialized::es6::ast::Walker;
+
+use std::io::{Error, ErrorKind};
+use std::path::Path;
+
+pub struct Options<'a> {
+    pub parser: &'a Shift,
+    pub lazification: u32,
+    pub quiet: bool,
+    pub depth: usize,
+    pub width: usize,
+    pub threshold: FilesContaining,
+}
+
+macro_rules! progress {
+    ($quiet:expr, $($args:tt)*) => {
+        if !$quiet {
+            println!($($args)*);
+        }
+    }
+}
+
+fn handle_path<'a>(
+    options: &mut Options<'a>,
+    shared_builder: &mut DictionaryBuilder,
+    shared_number_of_files: &mut usize,
+    source_path: &Path,
+    sub_dir: &Path,
+) -> Result<(), std::io::Error> {
+    progress!(options.quiet, "Treating {:?} ({:?})", source_path, sub_dir);
+    let is_dir = std::fs::metadata(source_path).unwrap().is_dir();
+    if is_dir {
+        let file_name = source_path
+            .file_name()
+            .ok_or_else(|| Error::new(ErrorKind::NotFound, "Cannot get file name from directory"))?;
+
+        let sub_dir = sub_dir.join(file_name);
+        for entry in std::fs::read_dir(source_path)?
+            .map(|dir| dir.unwrap())
+        {
+            handle_path(
+                options,
+                shared_builder,
+                shared_number_of_files,
+                &entry.path().as_path(),
+                &sub_dir,
+            )?;
+        }
+        return Ok(());
+    }
+    if let Some(Some("js")) = source_path.extension().map(std::ffi::OsStr::to_str) {
+        // Proceed
+    } else {
+        progress!(options.quiet, "Skipping {:?}", source_path);
+        return Ok(());
+    }
+
+    progress!(options.quiet, "Parsing.");
+
+    handle_path_or_text(options, shared_builder, shared_number_of_files, source_path)
+}
+
+fn handle_path_or_text<'a>(
+    options: &mut Options<'a>,
+    dictionary_builder: &mut DictionaryBuilder,
+    shared_number_of_files: &mut usize,
```

Comment:
Given that this is always incremented on success, perhaps worth removing a param here and keep track of count in the caller instead?